### PR TITLE
[tune] Try to recover from latest available checkpoint

### LINKED ISSUE #27790: [Tune] FileNotFoundError when resuming from Checkpoint
### What happened + What you expected to happen

@xwjiang2010 
This issue is related to my post here: [discuss.ray.io](https://discuss.ray.io/t/filenotfounderror-when-resuming-from-checkpoint/7146?u=stefan-1313). You asked me to tag you here so here you go :).

**What goes wrong:**
When running a Ray Tune experiment in a Docker container, stopping the container, and restarting the constrainer, the checkpoint cannot be found and the run crashes. It also happens on Windows when I force stop the Python process (e.g. kill from a task manager).

The issue only occurs when I set the `tune.run()` argument `keep_checkpoints_num` to something other than `None`.

I can 100% reproduce this issue. From a docker container it is easier to reproduce because you can easily stop the container and everything inside it. I can also reproduce on Windows, **but I really have to kill the Python process**, meaning not letting it gracefully stop (so e.g. in PyCharm, hitting the red stop button multiple times quickly to really force stop Python (just pressing the red stop button once wil gracefully stop the Python Process and the issue will not reproduce)).

**Reproduction steps.**
Set `keep_checkpoints_num=2` .
I start the Docker container:

1. [This point is in practice not really a problem, but it may 'give a hint' / 'be related to' the other 2 points] If I kill the docker container before more than 2 iterations have been completed and then restart the container, resume always works, _although sometimes Tune start from te beginning!_ (It is easier to set `keep_checkpoints_num=5` and you will see the same when no more than 5 iterations may have been completed.)
Set `keep_checkpoints_num=2` again for the remainder...
2. Suppose I let a trail run for 4 iterations, if I kill the docker container, minimum 5 seconds after the 4th iteration finished (based on a print from callback log_trial_result) and then, restart the container, resume work as expected, and the first iteration printed is 5 (based on a print from callback log_trial_result) .
3. Suppose I let a trail run for 4 iterations, if I kill the docker container, maximum 1 seconds after the 4th iteration finished (based on a print from callback log_trial_result) and then, restart the container, resume fails with the above error. It tries to resume from the 2th iteration (sometimes the 3th), and it may be deleted already by `keep_checkpoints_num`. If I inspect the output folder I can see the checkpoint folder checkpoint_000004 exists, and the 2th (or sometimes the 3th) was already deleted because of `keep_checkpoints_num=2`.
Off course, you can encounter cases where the 2th checkpoint was not deleted (as it is just luck which checkpoints score such that `keep_checkpoints_num` keeps them). If you wait longer and run for more than 4 iterations, more checkpoints have been deleted and you increase the probability of Tune trying to resume from a deleted checkpoint.

**A detail.**
I can reproduce this 100% but the timing described above in point 2 and 3 is required. This suggest the probability of getting this error is small as termination then always has to happen max a few seconds after completing an iteration.
I use the Azure Machine Learning Service and when running there, my Docker containers get stopped once in a while because I run them on low-priority nodes.
Unfortunately, I almost always get this error `FileNotFoundError` when resume on Azure is attempted. I assume it is because of the same issue as described in this post. However, I cannot explain why the probability of this occurring on my Azure nodes seems to be much larger.

**My best guess regarding the cause.**
It think Tune does NOT look at which checkpoint folders are present when determining from which checkpoint to resume. I think this information is stored in another file.
But writing to this file might be buffered. When the Python process (or Docker container running Python) is abruptly stopped, the buffer is not flushed and Tune attempts to resume from a wrong checkpoint. Unfortunately, this wrong checkpoint has been deleted when the new checkpoint was created.
_So maybe force flushing this buffer before deleting the old checkpoint?_
This is just a guess, I don't have any references to the source code to back this up.

**Terminal output for point 2 and 3.**
After point 2, I get this kind of output (everything works as expected):
The container is restarted when you see the line `Start Ray Tune.` somewhere halfway. This `*** SIGTERM received at time=1660223520 on cpu 10 ***` and everything after until restart is because of stopping the Docker container.
```
Start Ray Tune.
2022-08-11 13:11:34,633 INFO trial_runner.py:510 -- No local checkpoint was found. Ray Tune will now start a new experiment.
2022-08-11 13:11:34,791 ERROR syncer.py:147 -- Log sync requires rsync to be installed.
(SimpleQTrainer pid=516) 2022-08-11 13:11:36,935        INFO simple_q.py:187 -- In multi-agent mode, policies will be optimized sequentially by the multi-GPU optimizer. Consider setting `simple_optimizer=True` if this doesn't work for you.
(SimpleQTrainer pid=516) 2022-08-11 13:11:36,935        INFO trainer.py:903 -- Current log_level is ERROR. For more information, set 'log_level': 'INFO' / 'DEBUG' or use the -v and -vv flags.
Trial: SimpleQTrainer_LunarLander-v2_1ff52_00000 | Training iteration 1 | Episode reward mean: -167.08
Trial: SimpleQTrainer_LunarLander-v2_1ff52_00000 | Training iteration 2 | Episode reward mean: -177.17
Trial: SimpleQTrainer_LunarLander-v2_1ff52_00000 | Training iteration 3 | Episode reward mean: -161.51
Trial: SimpleQTrainer_LunarLander-v2_1ff52_00000 | Training iteration 4 | Episode reward mean: -145.84
*** SIGTERM received at time=1660223520 on cpu 10 ***
PC: @     0x7fb7a8c9aad8  (unknown)  pthread_cond_timedwait@@GLIBC_2.3.2
    @     0x7fb7a8c9f140  (unknown)  (unknown)
    @ ... and at least 1 more frames
[2022-08-11 13:12:00,221 E 1 1] logging.cc:325: *** SIGTERM received at time=1660223520 on cpu 10 ***
[2022-08-11 13:12:00,221 E 1 1] logging.cc:325: PC: @     0x7fb7a8c9aad8  (unknown)  pthread_cond_timedwait@@GLIBC_2.3.2
[2022-08-11 13:12:00,221 E 1 1] logging.cc:325:     @     0x7fb7a8c9f140  (unknown)  (unknown)
[2022-08-11 13:12:00,221 E 1 1] logging.cc:325:     @ ... and at least 1 more frames
Exception ignored in: <function RolloutWorker.__del__ at 0x7fb77e97d310>
Traceback (most recent call last):
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/util/tracing/tracing_helper.py", line 461, in _resume_span
TypeError: 'NoneType' object is not callable

Start Ray Tune.
2022-08-11 13:12:17,457 INFO trial_runner.py:515 -- A local experiment checkpoint was found and will be used to restore the previous experiment state.
2022-08-11 13:12:17,458 WARNING trial_runner.py:644 -- Attempting to resume experiment from /root/model_output/RayTesting_002. This will ignore any new changes to the specification.
2022-08-11 13:12:17,468 INFO tune.py:647 -- TrialRunner resumed, ignoring new add_experiment but updating trial resources.
2022-08-11 13:12:17,627 ERROR syncer.py:147 -- Log sync requires rsync to be installed.
(SimpleQTrainer pid=523) 2022-08-11 13:12:19,768        INFO simple_q.py:187 -- In multi-agent mode, policies will be optimized sequentially by the multi-GPU optimizer. Consider setting `simple_optimizer=True` if this doesn't work for you.
(SimpleQTrainer pid=523) 2022-08-11 13:12:19,768        INFO trainer.py:903 -- Current log_level is ERROR. For more information, set 'log_level': 'INFO' / 'DEBUG' or use the -v and -vv flags.
(SimpleQTrainer pid=523) 2022-08-11 13:12:19,818        INFO trainable.py:588 -- Restored on 172.17.0.2 from checkpoint: /root/model_output/SimpleQTrainer_LunarLander-v2_1ff52_00000_0_2022-08-11_13-11-34/tmpc6nwcvm2restore_from_object/checkpoint-4
(SimpleQTrainer pid=523) 2022-08-11 13:12:19,818        INFO trainable.py:597 -- Current state after restoring: {'_iteration': 4, '_timesteps_total': None, '_time_total': 17.697097063064575, '_episodes_total': 16}
Trial: SimpleQTrainer_LunarLander-v2_1ff52_00000 | Training iteration 5 | Episode reward mean: -145.32
Trial: SimpleQTrainer_LunarLander-v2_1ff52_00000 | Training iteration 6 | Episode reward mean: -153.24
Trial: SimpleQTrainer_LunarLander-v2_1ff52_00000 | Training iteration 7 | Episode reward mean: -168.02
<-- CUT - All the way until completion without erros... -->
```

After point 3, I get this kind of output (you see the `FileNotFoundError: [Errno Path does not exist]` error):
The container is restarted when you see the line `Start Ray Tune.` somewhere halfway.  This `*** SIGTERM received at time=1660223333 on cpu 8 ***` and everything after until restart is because of stopping the Docker container.
```
Start Ray Tune.
2022-08-11 13:08:32,604 INFO trial_runner.py:510 -- No local checkpoint was found. Ray Tune will now start a new experiment.
2022-08-11 13:08:32,763 ERROR syncer.py:147 -- Log sync requires rsync to be installed.
(SimpleQTrainer pid=522) 2022-08-11 13:08:34,898        INFO simple_q.py:187 -- In multi-agent mode, policies will be optimized sequentially by the multi-GPU optimizer. Consider setting `simple_optimizer=True` if this doesn't work for you.
(SimpleQTrainer pid=522) 2022-08-11 13:08:34,898        INFO trainer.py:903 -- Current log_level is ERROR. For more information, set 'log_level': 'INFO' / 'DEBUG' or use the -v and -vv flags.
Trial: SimpleQTrainer_LunarLander-v2_b375b_00000 | Training iteration 1 | Episode reward mean: -213.88
Trial: SimpleQTrainer_LunarLander-v2_b375b_00000 | Training iteration 2 | Episode reward mean: -217.36
Trial: SimpleQTrainer_LunarLander-v2_b375b_00000 | Training iteration 3 | Episode reward mean: -194.35
Trial: SimpleQTrainer_LunarLander-v2_b375b_00000 | Training iteration 4 | Episode reward mean: -190.43
*** SIGTERM received at time=1660223333 on cpu 8 ***
PC: @     0x7f3a9e681ad8  (unknown)  pthread_cond_timedwait@@GLIBC_2.3.2
    @     0x7f3a9e686140  (unknown)  (unknown)
    @ ... and at least 1 more frames
[2022-08-11 13:08:53,973 E 1 1] logging.cc:325: *** SIGTERM received at time=1660223333 on cpu 8 ***
[2022-08-11 13:08:53,973 E 1 1] logging.cc:325: PC: @     0x7f3a9e681ad8  (unknown)  pthread_cond_timedwait@@GLIBC_2.3.2
[2022-08-11 13:08:53,973 E 1 1] logging.cc:325:     @     0x7f3a9e686140  (unknown)  (unknown)
[2022-08-11 13:08:53,973 E 1 1] logging.cc:325:     @ ... and at least 1 more frames
Exception ignored in: <function RolloutWorker.__del__ at 0x7f3a74364310>
Traceback (most recent call last):
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/util/tracing/tracing_helper.py", line 461, in _resume_span
TypeError: 'NoneType' object is not callable

Start Ray Tune.
2022-08-11 13:09:08,819 INFO trial_runner.py:515 -- A local experiment checkpoint was found and will be used to restore the previous experiment state.
2022-08-11 13:09:08,820 WARNING trial_runner.py:644 -- Attempting to resume experiment from /root/model_output/RayTesting_002. This will ignore any new changes to the specification.
2022-08-11 13:09:08,827 INFO tune.py:647 -- TrialRunner resumed, ignoring new add_experiment but updating trial resources.
2022-08-11 13:09:08,978 ERROR ray_trial_executor.py:533 -- Trial SimpleQTrainer_LunarLander-v2_b375b_00000: Unexpected error starting runner.
Traceback (most recent call last):
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/tune/ray_trial_executor.py", line 526, in start_trial
    return self._start_trial(trial)
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/tune/ray_trial_executor.py", line 433, in _start_trial
    self.restore(trial)
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/tune/ray_trial_executor.py", line 744, in restore
    obj = TrainableUtil.checkpoint_to_object(value)
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/tune/utils/trainable.py", line 103, in checkpoint_to_object
    data_dict = TrainableUtil.pickle_checkpoint(checkpoint_path)
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/tune/utils/trainable.py", line 83, in pickle_checkpoint
    checkpoint_dir = TrainableUtil.find_checkpoint_dir(checkpoint_path)
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/tune/utils/trainable.py", line 118, in find_checkpoint_dir
    raise FileNotFoundError("Path does not exist", checkpoint_path)
FileNotFoundError: [Errno Path does not exist] /root/model_output/RayTesting_002/SimpleQTrainer_LunarLander-v2_b375b_00000_0_2022-08-11_13-08-32/checkpoint_000002/checkpoint-2
(SimpleQTrainer pid=521) 2022-08-11 13:09:11,160        INFO simple_q.py:187 -- In multi-agent mode, policies will be optimized sequentially by the multi-GPU optimizer. Consider setting `simple_optimizer=True` if this doesn't work for you.
(SimpleQTrainer pid=521) 2022-08-11 13:09:11,160        INFO trainer.py:903 -- Current log_level is ERROR. For more information, set 'log_level': 'INFO' / 'DEBUG' or use the -v and -vv flags.
Traceback (most recent call last):
  File "/home/developer/workspace/main_MinWorkingExample.py", line 56, in <module>
    tune.run(SimpleQTrainer,
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/tune/tune.py", line 741, in run
    raise TuneError("Trials did not complete", incomplete_trials)
ray.tune.error.TuneError: ('Trials did not complete', [SimpleQTrainer_LunarLander-v2_b375b_00000])
Exception ignored in: <function RolloutWorker.__del__ at 0x7fe65b381310>
Traceback (most recent call last):
  File "/opt/conda/envs/cenv39ray/lib/python3.9/site-packages/ray/util/tracing/tracing_helper.py", line 461, in _resume_span
TypeError: 'NoneType' object is not callable
```

### Versions / Dependencies

I include a reproduction script including a docker and conda requirements file.
Basically Python 3.9 and Ray 1.13.0.

### Conda environment
I used this conda environment both on Windows and on Docker, just CPU for the sake of simplicity and size:
```
# conda_env_definition.yml

name: cenv39ray
channels:
- pytorch
- conda-forge
dependencies:
- python=3.9
- pytorch=1.11.0
- torchaudio
- torchvision
- cpuonly
- swig  # Needed to build 'box2d-py' in the pip install.
- pip=21.0.1
- pip:
   - ray==1.13.0
   - ray[tune]==1.13.0
   - ray[rllib]==1.13.0
   - box2d-py  # Because of using 'LunarLander-v2' environment.
   - psutil==5.8.0
   - gputil==1.4.0
```

### Docker container
Optional, build this docker image:
```
# dockerfile.txt

# Use base image having miniconda installed.
FROM continuumio/miniconda3

# For building pip package 'box2d-py' required because of using 'LunarLander-v2' environment.
RUN apt-get update && \
    apt-get install -y \
	gcc \
	g++

# Create environment variable for recognizing Docker container.
ENV RUNNING_IN_DOCKER_CONTAINER true

# Create conda environment.
COPY conda_env_definition.yml /data/conda_env_definition.yml
RUN conda env create --file=/data/conda_env_definition.yml
```

Command to build the Docker image:
`docker build -t ray_checkpoint_issue -f dockerfile.txt .`

Command to create the Docker container and start it (run from Powershell):
(Note, you need to change the volume path, I used the WSL-filesystem for the output folder (where the checkpoints are stored) to more closely resemble the situation on Azure, but this is not required.)
```
docker run -i -t `
    --name ray_checkpoint_issue `
	-v 'D:\dev\ray_checkpoint_issue:/home/developer/workspace' `
	-v '\\wsl$\Ubuntu-20.04\home\stefan\model_output:/root/model_output' `
	-w /home/developer/workspace `
	--shm-size=11gb `
	ray_checkpoint_issue `
	bash -c 'source /opt/conda/etc/profile.d/conda.sh && conda activate cenv39ray && python -u main_MinWorkingExample.py'
```

Then, from Docker Desktop, you can stop/start the container from the 'containers' menu. When starting from Docker Desktop, also the Python code is automatically started again, just like Azure Machine Learning would do after a low-priority stop event.

### Python code
Minimum working example with only vanilla Ray blocks.
_Note, I messed around with the learning/exploration config to make sure this model does not train very wel! So that it is random if the next checkpoint performs better or worse than the previous. If you model trains too wel, last checkpoints are also the best and therefore only the last `keep_checkpoints_num` are kept, then the issue does not reproduce very well._

_Also, I increased `train_batch_size` from 32 to 1024 so a network update takes some time. This gives you the time to wait a wile before stopping the Python process, else, while you wait, already a new iterations is finished which interferes with the reproduction of the issue. If you system is much faster then mine, increase it further e.g. to 2048._

```
# Imports.
import os
import ray
from ray import tune
from ray.rllib.agents.dqn.simple_q import DEFAULT_CONFIG, SimpleQTrainer
from ray.rllib.utils.typing import TensorType
from ray.tune.logger import LoggerCallback


# Define a callback to print a line everytime an iteration is finished.
class TrailIterationProgressPrinterCallback(LoggerCallback):
    def log_trial_result(self, iteration: int, trial: "Trial", result: dict):
        print(f"Trial: {trial} | Training iteration {result['training_iteration']} | Episode reward mean: {result['episode_reward_mean']:.2f}")


# Configure checkpointing.
experiment_name = "RayTesting_002"
resume = "AUTO"
if os.getenv('RUNNING_IN_DOCKER_CONTAINER'):  # If running in a docker container, use a different path.
    local_dir = r'/root/model_output'
else:
    local_dir = r'D:\dev\ray_checkpoint_issue\Output'


# Create experiment config.
number_of_iterations = 20
env_name = "LunarLander-v2"

config_simple = DEFAULT_CONFIG.copy()
config_simple["env"] = env_name
config_simple["framework"] = "torch"
config_simple["_disable_preprocessor_api"] = True
config_simple["num_gpus"] = 0
config_simple["num_workers"] = 0
config_simple["lr"] = 0.00001
config_simple["rollout_fragment_length"] = 1
config_simple["train_batch_size"] = 1024  # Large, so training after an iteration takes some time and you can actually wait a while before stopping the Python process running this script.
config_simple["timesteps_per_iteration"] = 250
config_simple["target_network_update_freq"] = 500  # Default.
config_simple["log_level"] = "ERROR"
config_simple["explore"] = True
config_simple["exploration_config"] = {
    "type": ray.rllib.utils.exploration.epsilon_greedy.EpsilonGreedy,
    "initial_epsilon": 1.0,
    "final_epsilon": 0.99,  # use a random agent, so we get random checkpoints which perform better and are therefore kept.
    "epsilon_timesteps": config_simple["timesteps_per_iteration"] * number_of_iterations * 0.7,
    }

# Start tune and run experiments.
print('')
print('Start Ray Tune.')
tune.run(SimpleQTrainer,
         config=config_simple,
         callbacks=[TrailIterationProgressPrinterCallback()],
         stop={"training_iteration": number_of_iterations},
         verbose=0,
         resume=resume,
         checkpoint_freq=1,
         keep_checkpoints_num=2,
         checkpoint_score_attr='episode_reward_mean',
         local_dir=local_dir,
         name=experiment_name,
         )

```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Tune tracks the latest available checkpoint per trial in the trial metadata, which is stored on the driver. This data is periodically synced to cloud storage, but can get out of sync if e.g. the node where the driver was on is terminated.

When experiments are continued, this can then lead to us not finding the checkpoints that tune wants to recover from (e.g. because we only keep a certain number of checkpoints). The current fallback behavior is to start the trials from scratch. However, instead we would usually like to restart the trial from the latest available checkpoint instead.

This PR adds fully unit- and integration-tested functionality for this, and guards it behind a feature env flag, that is enabled per default.

## Related issue number

Closes #27790

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tests/test_tuner_restore.py [hunk 1]
```diff
@@ -352,6 +354,62 @@ def test_tuner_restore_from_cloud(ray_start_2_cpus, tmpdir):
     tuner3.fit()
 
 
+@pytest.mark.parametrize(
+    "upload_uri",
+    [None, "memory:///test/test_tuner_restore_latest_available_checkpoint"],
+)
+def test_tuner_restore_latest_available_checkpoint(
```
Comment (xwjiang2010) [inline python/ray/tune/tests/test_tuner_restore.py:361]: you may need to move this test before the resource exhaustion one... See comment there on that test.

File: python/ray/tune/trainable/trainable.py [hunk 2]
```diff
@@ -650,6 +710,17 @@ def restore(
                 checkpoint.to_directory(checkpoint_path)
 
         if not os.path.exists(checkpoint_path):
+            if fallback_to_latest:
+                logger.info(
+                    f"Checkpoint path was not available, trying to recover from latest "
+                    f"available checkpoint instead. Unavailable checkpoint path: "
+                    f"{checkpoint_path}"
+                )
+                checkpoint_path = self._get_latest_available_checkpoint()
```
Comment (xwjiang2010) [inline python/ray/tune/trainable/trainable.py:719]: logging on what path we are using instead?

**Comment by xwjiang2010** [comment]
Do we purge all non-relevant checkpoints when a trial is failed to restore even after several retries?
I checked the code and it doesn't seem that we physically delete the folders either locally or in cloud storage.
would that cause an issue?

say max_failures = 3
1st run logs iteration 1, 2, 3, 4, 5, 6, 7
restore happens and fails
2nd run logs iteration 1, 2,
try to restore again and this time somehow fail to find the checkpoint and restore from latest, which is 7 (from 1st run)

would that cause issue?

**Comment by krfricke** [comment]
So we do override existing checkpoint directories when we log a checkpoint with the same step index. I see what you mean, and no, we don't purge existing checkpoints.

I don't think this is a big issue here - if the trial restore fails on the first try (when we want to restore from 7), then trains from scratch until 2, then fails, and then recovers from 7, we end up with behavior we originally wanted. 

I think it's fine to keep this for now, but we can revisit this if any problems come up.

File: python/ray/tune/trainable/trainable.py [hunk 3]
```diff
@@ -502,6 +503,56 @@ def save(
...
+        latest_local_checkpoint = self._get_latest_local_available_checkpoint()
+        latest_remote_checkpoint = self._get_latest_remote_available_checkpoint()
+
+        if not latest_local_checkpoint:
+            return latest_remote_checkpoint
+        elif not latest_remote_checkpoint:
+            return latest_local_checkpoint
+
+        # Else, both are available
+        return max([latest_local_checkpoint, latest_remote_checkpoint])
```
Comment (xwjiang2010) [inline python/ray/tune/trainable/trainable.py:516]: is this comparing the absolute path between local and remote checkpoints?

Comment (xwjiang2010) [reply python/ray/tune/trainable/trainable.py:516]: shouldn't we compare the iteration number instead?

**Review by xwjiang2010 — APPROVED** [review]

File: python/ray/tune/trainable/trainable.py (hunk 3 above)
Comment (krfricke) [reply python/ray/tune/trainable/trainable.py:516]: It should be fine as we do `os.path.join(self._logdir, name)` in both cases!

**Pull request merged**